src/engine/Match_tainting_mode.ml

```diff
@@ -447,42 +448,80 @@ let rec convert_taint_call_trace = function
           call_trace = convert_taint_call_trace ct;
         }
 
-let taint_trace_of_src_to_sink source tokens sink =
-  {
-    Pattern_match.source = convert_taint_call_trace source.T.call_trace;
-    tokens;
-    sink = convert_taint_call_trace sink;
-  }
+let taint_trace_of_src_traces_and_sink sources sink =
+  { Pattern_match.sources; sink = convert_taint_call_trace sink }
 
-let pm_of_finding finding =
+let pms_of_finding finding =
   match finding with
-  | T.SrcToSink { source; tokens; sink; merged_env } ->
-      (* We always report the finding on the sink that gets tainted, the call trace
-       * must be used to explain how exactly the taint gets there. At some point
-       * we experimented with reporting the match on the `sink`'s function call that
-       * leads to the actual sink. E.g.:
-       *
-       *     def f(x):
-       *       sink(x)
-       *
-       *     def g():
-       *       f(source)
-       *
-       * Here we tried reporting the match on `f(source)` as "the line to blame"
-       * for the injection bug... but most users seem to be confused about this. They
-       * already expect Semgrep (and DeepSemgrep) to report the match on `sink(x)`.
-       *)
-      let taint_trace =
-        Some (lazy (taint_trace_of_src_to_sink source tokens sink))
-      in
-      let sink_pm, _ = T.pm_of_trace sink in
-      Some { sink_pm with env = merged_env; taint_trace }
-  | T.SrcToReturn _
-  (* TODO: We might want to report functions that let input taint
-   * go into a sink (?) *)
-  | T.ArgToSink _
-  | T.ArgToReturn _ ->
-      None
+  | T.ToReturn (_taints, _) ->
+      (* TODO: We might want to report functions that let input taint
+         * go into a sink (?) *)
+      []
+  | ToSink { taints_with_precondition = taints, requires; sink; merged_env } ->
+      if not (D.taints_satisfy_requires (T.Taint_set.of_list taints) requires)
+      then []
+      else
+        (* these arg taints are not useful to us, because we are within
+           the function, not at the call-site. so we don't know what
+           the argument taints are.
+        *)
+        let default_taints, labeled_source_taints, _args_taints =
+          taints
+          |> Common.partition_either3 (fun { T.orig; tokens } ->
+                 match orig with
+                 | Src src when src.label =*= Rule.default_source_label ->
+                     Left3 (src, tokens)
+                 | Src src -> Middle3 (src, tokens)
+                 | Arg arg -> Right3 arg)
+        in
+        (* The old behavior used to be that, for sinks with a `requires`, we would
+           generate a finding per every single taint source going in. Later deduplication
+           would deal with it.
+           We will continue that behavior for taint sinks with no `requires`, but for multiple
```

Comment:
Haven't you done the opposite? For sinks with no `requires:` this generates one finding, and for sinks with `requires:` it generates one finding per sink.